File: interpreters/fsharp/Command.fs

module Command

type Command =
  | IncPtr
  | DecPtr
  | IncVal
  | DecVal
  | PutChar
  | GetChar
  | OpenBr
  | CloseBr

let parseCode code =
  code
  |> List.choose (function
      | '>' -> Some IncPtr
      | '<' -> Some DecPtr
      | '+' -> Some IncVal
      | '-' -> Some DecVal
      | '.' -> Some PutChar
      | ',' -> Some GetChar
      | '[' -> Some OpenBr
      | ']' -> Some CloseBr
      | _ -> None)

let buildJumpTable (program: Command list) =
  let rec loop i stack map =
    match program |> List.tryItem i with
    | None -> map // "Done" condition
    | Some OpenBr -> loop (i + 1) (i :: stack) map
    | Some CloseBr ->
        match stack with
        | [] -> failwithf "Error: unmatched ']' at %d" i
        | openIdx :: rest ->
            loop (i + 1) rest (map |> Map.add openIdx i |> Map.add i openIdx)
    | Some _ -> loop (i + 1) stack map
  loop 0 [] Map.empty

File: interpreters/haskell/app/Command.hs
module Command(Command, parseCmds, buildJumpTable) where

import qualified Data.Map as Map

data Command = IncPtr
  | DecPtr
  | IncVal
  | DecVal
  | OpenBr
  | CloseBr
  | PutChar
  | GetChar
  deriving Show

parseCmds :: [Char] -> [Command]
parseCmds [] = []
parseCmds (ch : rest) = case ch of
  '+' -> IncVal : parseCmds rest
  '-' -> DecVal : parseCmds rest
  '>' -> IncPtr : parseCmds rest
  '<' -> DecPtr : parseCmds rest
  '[' -> OpenBr : parseCmds rest
  ']' -> CloseBr : parseCmds rest
  '.' -> PutChar : parseCmds rest
  ',' -> GetChar : parseCmds rest
  _ -> parseCmds rest


buildJumpTable :: [Command] -> Map.Map Int Int
buildJumpTable cmds = buildJumpTable' cmds [] 0

buildJumpTable' :: [Command] -> [Int] -> Int -> Map.Map Int Int
buildJumpTable' [] _ _ = Map.empty
buildJumpTable' (next : rest) stack count = case next of
  OpenBr -> buildJumpTable' rest (count : stack) (count+1)
  CloseBr -> do
    let stackH : stackRest = stack
    Map.insert stackH count (Map.insert count stackH (buildJumpTable' rest stackRest (count+1)))
  _ -> buildJumpTable' rest stack (count+1)


